lib/src/workspace.rs

```diff
@@ -461,11 +464,64 @@ pub struct LockedWorkspace<'a> {
     locked_wc: Box<dyn LockedWorkingCopy>,
 }
 
+#[derive(Debug, Error)]
+pub enum RecoverWorkspaceError {
+    #[error(transparent)]
+    Backend(#[from] BackendError),
+    #[error(transparent)]
+    Reset(#[from] ResetError),
+    #[error("Cannot rewrite the root commit")]
+    RewriteRootCommit,
+}
+
+impl From<RewriteRootCommit> for RecoverWorkspaceError {
+    fn from(_value: RewriteRootCommit) -> Self {
+        Self::RewriteRootCommit
+    }
+}
+
 impl LockedWorkspace<'_> {
     pub fn locked_wc(&mut self) -> &mut dyn LockedWorkingCopy {
         self.locked_wc.as_mut()
     }
 
+    // Recover this workspace to its last known checkout.
+    pub fn create_and_check_out_recovery_commit(
```

Comment:
It's odd that `LockedWorkspace` has a method that does very specific thing. Maybe this can be moved to cli_util?

Reply:
I'd prefer to keep this in lib if possible for use by extensions.  How about moving it to working_copy.rs as a free function?

Reply:
A free function sounds good to me, but can you add some commit description about why it should be in the lib? Ah, "the extensions" means extension commands/UI that uses jj-lib?

Also, iirc, we generally avoid adding commit description (that might have to be translated) to lib, so `"RECOVERY COMMIT FROM ...` might have to be a function parameter.

Reply:
Done, yes, and done.